### PR TITLE
CalorimeterHitReco: use TCling to define SF via expressions

### PR DESCRIPTION
**Pull request by veprbl**
### Briefly, what does this PR introduce?

This introduces an ability to define sampling fraction that varies across the calorimeter cells.
This should enable following kind of studies for TDR:
 - Non-uniform tower gains (including dead channels)
 Example: `-PEEMC:EcalEndcapNRecHits:samplingFraction='[row, column](){ static map<std::tuple<double, double>, double> sf; auto index = std::make_tuple(row, column); if (!sf.count(index)) { std::default_random_engine gen(row + 100 * column); std::normal_distribution R{1.0, 0.1}; sf[index] = R(gen); } return sf[index]; }()'` (yes, the static variables in lambda are actually static in the scope it was defined in, so this provides caching)
 - Rapidity-dependent calibrations

Use of TCling is very ugly and inefficient, this is not mean to be permanent (... they always say). We should be replacing this with some proper database, or a better scripting language.

Closes: #799
Resolves: #1337

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No

### Does this PR change default behavior?
No

### CODE REVIEW

File: src/algorithms/calorimetry/CalorimeterHitReco.cc [hunk 1]
```diff
@@ -39,7 +38,7 @@ using namespace dd4hep;
 
 namespace eicrecon {
 
-void CalorimeterHitReco::init() {
+void CalorimeterHitReco::init(std::shared_ptr<Evaluator_service> evaluator) {
```
Comment (wdconinc) [inline src/algorithms/calorimetry/CalorimeterHitReco.cc:41]: Can we avoid this? We are (were) moving towards a uniform interface for all algorithms: `void init() final` and `void process(const Input&, const Output&) const final`, so we don't need to rewrite all these factories all the time. Is there a way that we can pass this evaluator service through other means?

@sly2j for potential `algorithms::` interface ideas.

Comment (veprbl) [reply src/algorithms/calorimetry/CalorimeterHitReco.cc:41]: Can we register JANA service as algorithms service?

File: src/algorithms/calorimetry/CalorimeterHitReco.h [hunk 2]
```diff
@@ -47,15 +51,20 @@ namespace eicrecon {
                             {"outputRecHitCollection"},
                             "Reconstruct hit from digitized input."} {}
 
-    void init() final;
+    void init();
     void process(const Input&, const Output&) const final;
 
   private:
 
+    algorithms::Service<EvaluatorSvc> m_evaluatorSvc {""};
```
Comment (veprbl) [inline src/algorithms/calorimetry/CalorimeterHitReco.h:59]: This is probably incorrect.

File: src/algorithms/calorimetry/CalorimeterHitRecoConfig.h [hunk 3]
```diff
@@ -23,8 +23,7 @@ namespace eicrecon {
     double                   thresholdValue{0};
 
     // sampling fraction
-    double                   sampFrac{1.0};
-    std::vector<double>      sampFracLayer{};
+    std::string              sampFrac{"1.0"};
```
Comment (wdconinc) [inline src/algorithms/calorimetry/CalorimeterHitRecoConfig.h:26]: Can we handle this as follows?
```suggestion
    std::variant<double,std::string> sampFrac{1.0};
```

Main benefit would actually be for expansion of this pattern. If we start to roll this out as everywhere, but it it requires us to change defaults to runtime-evaluated strings, we lose the compile-time validity checks on numerical values.

File: src/detectors/FHCAL/FHCAL.cc [hunk 4]
```diff
@@ -156,23 +156,7 @@ extern "C" {
...
-              0.037, //  6
-              0.037, //  7
-              0.037, //  8
-              0.037, //  9
-              0.037, // 10
-              0.037, // 11
-              0.037, // 12
-              0.037, // 13
-            },
+            .sampFrac = "(rlayerz == 0) ? 0.019 : 0.037", // 0.019 only in the 0-th tile
```
Comment (wdconinc) [inline src/detectors/FHCAL/FHCAL.cc:159]: Nice :-)

File: src/factories/calorimetry/CalorimeterHitReco_factory.h [hunk 5]
```diff
@@ -28,7 +28,7 @@ class CalorimeterHitReco_factory : public JOmniFactory<CalorimeterHitReco_factor
     ParameterRef<double> m_resolutionTDC {this, "resolutionTDC", config().resolutionTDC};
     ParameterRef<double> m_thresholdFactor {this, "thresholdFactor", config().thresholdFactor};
     ParameterRef<double> m_thresholdValue {this, "thresholdValue", config().thresholdValue};
-    ParameterRef<double> m_samplingFraction {this, "samplingFraction", config().sampFrac};
+    ParameterRef<std::string> m_samplingFraction {this, "samplingFraction", config().sampFrac};
```
Comment (wdconinc) [inline src/factories/calorimetry/CalorimeterHitReco_factory.h:31]: Though this probably doesn't handle variants... Maybe we can somehow 'pick one' here, naturally the string. Maybe:
```
    ParameterRef<std::string> m_samplingFraction {this, "samplingFraction", std::get<std::string>(config().sampFrac)};
```

File: src/algorithms/calorimetry/CalorimeterHitReco.h [hunk 6]
```diff
@@ -47,7 +49,7 @@ namespace eicrecon {
                             {"outputRecHitCollection"},
                             "Reconstruct hit from digitized input."} {}
 
-    void init() final;
+    void init();
```
Comment (wdconinc) [inline src/algorithms/calorimetry/CalorimeterHitReco.h:52]: Can still be `final`.

File: src/services/evaluator/EvaluatorSvc.cc [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#include "EvaluatorSvc.h"
+
+namespace eicrecon {
+
+void EvaluatorSvc::init() {
+}
+
+std::function<double(const std::unordered_map<std::string, double>&)>
+EvaluatorSvc::_compile(const std::string& expr, std::vector<std::string> params) {
+  std::string func_name = fmt::format("_eicrecon_{}", function_id++);
```
Comment (wdconinc) [inline src/services/evaluator/EvaluatorSvc.cc:20]:
```
std::lock_guard<std::mutex> lock(m_mutex);
```

File: src/algorithms/calorimetry/CalorimeterHitReco.h (hunk 6 above)
Comment (veprbl) [reply src/algorithms/calorimetry/CalorimeterHitReco.h]: Nice catch

File: src/services/evaluator/EvaluatorSvc.cc (hunk 7 above)
Comment (veprbl) [reply src/services/evaluator/EvaluatorSvc.cc]: Done

**Comment by veprbl** [comment]
@ruse-traveler Check this out. This is like adjacency matrix but for your calibration constants.

**Comment by ruse-traveler** [comment]
> @ruse-traveler Check this out. This is like adjacency matrix but for your calibration constants.

Nice! This is very cool! So to check my understanding: if I wanted to 0 out some channels, I would do something like the following, right?
```
.sampFrac = "(eta == phi) ? 0.0 : 0.033"  // unalive all channels on a diagonal, otherwise use default
```

(c.c. @FriederikeBock )

**Comment by veprbl** [comment]
Yes. That's the idea. The plan is to apply the same for CalorimeterHitDigi, and add additional parameters that can be used, including energy for a non-linear response.

File: src/algorithms/calorimetry/CalorimeterHitRecoConfig.h (hunk 3 above)
Comment (wdconinc) [reply src/algorithms/calorimetry/CalorimeterHitRecoConfig.h:26]: Just to make sure you don't think you need to keep this open on my behalf, I'm ok moving away from the variant approach if it doesn't work technically, but I am worried about what we lose when we stick everything in strings including constants.

**Comment by ruse-traveler** [comment]
> Yes. That's the idea. The plan is to apply the same for CalorimeterHitDigi, and add additional parameters that can be used, including energy for a non-linear response.

Gotcha! So then this is also where we would stick additional "noise" if we wanted something other than what's in the vanilla algorithm...

**Review by ruse-traveler — APPROVED** [review]
Though I do share Wouter's concern about defining constants in strings, I would be okay with integrating these changes as-is if the variant approach doesn't work. This Should address the two primary digitization needs from the perspective of the LFHCAL..

**Comment by ruse-traveler** [comment]
Checking in on this: aside from the conflicts, are you ready to merge this @veprbl ?

**Comment by veprbl** [comment]
> Checking in on this: aside from the conflicts, are you ready to merge this @veprbl ?

Yes. This needs another approve.

**Review by ruse-traveler — DISMISSED** [review]
👍

**Comment by veprbl** [comment]
@ruse-traveler One more time, please :)

**Review by ruse-traveler — APPROVED** [review]
👍 👍

**Pull request merged**